sdk/storage/src/clients/table_client.rs

```diff
@@ -0,0 +1,68 @@
+use crate::clients::TableServiceClient;
+use crate::table::requests::*;
+use azure_core::errors::AzureError;
+use http::method::Method;
+use http::request::{Builder, Request};
+use std::sync::Arc;
+
+pub trait AsTableClient<TN: Into<String>> {
+    fn as_table_client(&self, table_name: TN) -> Arc<TableClient>;
+}
+
+impl<TN: Into<String>> AsTableClient<TN> for Arc<TableServiceClient> {
+    fn as_table_client(&self, table_name: TN) -> Arc<TableClient> {
+        TableClient::new(self.clone(), table_name.into())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TableClient {
+    table_service_client: Arc<TableServiceClient>,
+    table_name: String,
+}
+
+impl TableClient {
+    pub(crate) fn new(
+        table_service_client: Arc<TableServiceClient>,
+        table_name: String,
+    ) -> Arc<Self> {
```

Comment:
Not sure how I feel about the `Arc` around the table client. The inner client is already an Arc. Is cloning the string that expensive? If so, should it also be wrapped in Arc to make the TableClient cheap to clone?

Reply:
Right now it does not make sense, aside for similarity to the blob apis (keep in mind I haven't touched the table API in this PR, I just wanted it to compile and behave as before).

In a future PR we should clean the Table APIs as well and that will lead to deeper hierarchy (that involves more reference counting).

For example `InsertEntity` has this "old style" signature:

```rust
pub fn insert_entity<'a, E>( 
    &'a self,
    entity: &'a E,
) -> InsertEntityBuilder<
    'a, 
    E,  
    crate::table::requests::RowKeyMissing,
    crate::table::requests::PartitionKeyMissing,
>   
where
    E: serde::Serialize,
{   
    InsertEntityBuilder::new(self, entity)
}   
```

I think we could end up with a `TablePartitionClient` that has this function: 

```rust
pub fn insert_entity<'a, E>( 
    &'a self,
    entity: &'a E,
    row_key: &RowKey,
) -> InsertEntityBuilder<'a, E> 
where
    E: serde::Serialize,
{   
    InsertEntityBuilder::new(self, entity)
}   
```

No gnerics, and a much simpler signature.